rpc/src/server.rs

```diff
@@ -104,3 +104,128 @@ impl RpcServer {
         }
     }
 }
+
+use crate::error::RPCError;
+use futures::{future::ok, future::Either, Async, Future, Poll};
+use jsonrpc_core::{middleware::Middleware, Call, Id, Metadata, Output, Request, Response};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+#[derive(Clone, Default)]
+pub struct ModuleEnableCheck(Arc<HashMap<String, Arc<String>>>);
+
+impl ModuleEnableCheck {
+    pub fn new(set: HashMap<String, Arc<String>>) -> Self {
+        ModuleEnableCheck(Arc::new(set))
+    }
+}
+
+/// Dummy future used as a noop result of middleware.
+pub type NoopFuture = Box<dyn Future<Item = Option<Response>, Error = ()> + Send>;
+/// Dummy future used as a noop call result of middleware.
+pub type NoopCallFuture = Box<dyn Future<Item = Option<Output>, Error = ()> + Send>;
+
+impl<M: Metadata> Middleware<M> for ModuleEnableCheck {
+    type Future = NoopFuture;
+    type CallFuture = NoopCallFuture;
+
+    fn on_request<F, X>(&self, request: Request, meta: M, next: F) -> Either<Self::Future, X>
+    where
+        F: Fn(Request, M) -> X + Send + Sync,
+        X: Future<Item = Option<Response>, Error = ()> + Send + 'static,
+    {
+        let output = |method: &jsonrpc_core::types::MethodCall, module: &str| {
+            Output::from(
+                Err(RPCError::custom(
+                    RPCError::Invalid,
+                    format!(
+                        "You need to enable `{module}` module to invoke `{method}` rpc, \
+                        please modify `rpc.modules` {miner_info} of configuration file ckb.toml and restart the ckb node",
+                        method = method.method, module = module, miner_info = if module == "Miner" {"and `block_assembler`"} else {""}
+                    ),
+                )),
+                method.id.clone(),
+                method.jsonrpc,
+            )
+        };
+        match request {
```

Comment:
Instead of adding a pre-middleware, it's better to implement it as a post-middleware. The middleware only checks whether the module is enabled when the handler returns `Error::method_not_found`.

Reply:
post-middleware is not a better solution, at least not here.

1. There is no request information in the response, which means you have to save(clone） a piece of request information for possible error return
2. Especially for batch requests, the complexity is doubled